File: src/android/GameAnalyticsCordova.java
package com.gameanalytics.sdk.cordova;

import com.gameanalytics.sdk.GameAnalytics;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class GameAnalyticsCordova extends CordovaPlugin
{
    private CordovaInterface mCordova;
    private static final String VERSION = "4.2.5";

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView)
    {
        super.initialize(cordova, webView);
        mCordova = cordova;
    }

    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) throws JSONException
    {
        if(action.equals("configureAvailableCustomDimensions01"))
        {
            JSONArray jsonArray = data.optJSONArray(0);

            ArrayList<String> array = new ArrayList<String>();

            if(jsonArray != null)
            {
                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }
            }

            GameAnalytics.configureAvailableCustomDimensions01(array.toArray(new String[0]));
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureAvailableCustomDimensions02"))
        {
            JSONArray jsonArray = data.optJSONArray(0);

            ArrayList<String> array = new ArrayList<String>();

            if(jsonArray != null)
            {
                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }
            }

            GameAnalytics.configureAvailableCustomDimensions02(array.toArray(new String[0]));
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureAvailableCustomDimensions03"))
        {
            JSONArray jsonArray = data.optJSONArray(0);

            ArrayList<String> array = new ArrayList<String>();

            if(jsonArray != null)
            {
                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }
            }

            GameAnalytics.configureAvailableCustomDimensions03(array.toArray(new String[0]));
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureAvailableResourceCurrencies"))
        {
            JSONArray jsonArray = data.optJSONArray(0);

            ArrayList<String> array = new ArrayList<String>();

            if(jsonArray != null)
            {
                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }
            }

            GameAnalytics.configureAvailableResourceCurrencies(array.toArray(new String[0]));
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureAvailableResourceItemTypes"))
        {
            JSONArray jsonArray = data.optJSONArray(0);

            ArrayList<String> array = new ArrayList<String>();

            if(jsonArray != null)
            {
                for(int i = 0; i < jsonArray.length(); ++i)
                {
                    array.add(jsonArray.optString(i, ""));
                }
            }

            GameAnalytics.configureAvailableResourceItemTypes(array.toArray(new String[0]));
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureBuild"))
        {
            String build = data.optString(0, "");
            GameAnalytics.configureBuild(build);
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureAutoDetectAppVersion"))
        {
            boolean flag = data.optBoolean(0, false);
            GameAnalytics.configureAutoDetectAppVersion(flag);
            callbackContext.success();
            return true;
        }
        else if(action.equals("configureUserId"))
        {
            String uId = data.optString(0, "");
            GameAnalytics.configureUserId(uId);
            callbackContext.success();
            return true;
        }
        else if(action.equals("initialize"))
        {
            JSONObject args = data.optJSONObject(0);
            String gameKey = "";
            String gameSecret = "";
            String sdkVersion = "cordova " + VERSION;
            //String sdkVersion = "";

            if(args != null)
            {
                gameKey = args.optString("gameKey", "");
                gameSecret = args.optString("secretKey", "");
                sdkVersion = args.optString("sdkVersion", "cordova " + VERSION);
                //sdkVersion = args.optString("sdkVersion", "");
            }

            GameAnalytics.configureSdkGameEngineVersion(sdkVersion);
            //GameAnalytics.configureGameEngineVersion("cordova " + VERSION);
            GameAnalytics.initialize(mCordova.getActivity(), gameKey, gameSecret);
            callbackContext.success();
            return true;
        }
        else if(action.equals("addBusinessEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            String currency = "";
            int amount = 0;
            String itemType = "";
            String itemId = "";
            String cartType = "";
            String receipt = "";
            String signature = "";
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                currency = args.optString("currency", "");
                amount = args.optInt("amount", 0);
                itemType = args.optString("itemType", "");
                itemId = args.optString("itemId", "");
                cartType = args.optString("cartType", "");
                receipt = args.optString("receipt", "");
                signature = args.optString("signature", "");
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            if(receipt.length() > 0 && signature.length() > 0)
            {
                GameAnalytics.addBusinessEvent(currency, amount, itemType, itemId, cartType, receipt, "google_play", signature,
                        fields, mergeFields);
            }
            else
            {
                GameAnalytics.addBusinessEvent(currency, amount, itemType, itemId, cartType, fields, mergeFields);
            }

            callbackContext.success();
            return true;
        }
        else if(action.equals("addResourceEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            int flowType = 0;
            String currency = "";
            float amount = 0;
            String itemType = "";
            String itemId = "";
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                flowType = args.optInt("flowType", 0);
                currency = args.optString("currency", "");
                amount = (float)args.optDouble("amount", 0);
                itemType = args.optString("itemType", "");
                itemId = args.optString("itemId", "");
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            GameAnalytics.addResourceEvent(flowType, currency, amount, itemType, itemId, fields, mergeFields);
            callbackContext.success();
            return true;
        }
        else if(action.equals("addProgressionEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            int progressionStatus = 0;
            String progression01 = "";
            String progression02 = "";
            String progression03 = "";
            double score = 0;
            boolean sendScore = false;
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                progressionStatus = args.optInt("progressionStatus", 0);
                progression01 = args.optString("progression01", "");
                progression02 = args.optString("progression02", "");
                progression03 = args.optString("progression03", "");
                score = args.optInt("score", 0);
                sendScore = args.has("score");
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            if(sendScore)
            {
                GameAnalytics.addProgressionEvent(progressionStatus, progression01, progression02, progression03, score,
                        fields, mergeFields);
            }
            else
            {
                GameAnalytics.addProgressionEvent(progressionStatus, progression01, progression02, progression03,
                        fields, mergeFields);
            }

            callbackContext.success();
            return true;
        }
        else if(action.equals("addDesignEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            String eventId = "";
            double value = 0;
            boolean sendValue = false;
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                eventId = args.optString("eventId", "");
                value = args.optDouble("value", 0);
                sendValue = args.has("value");
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            if(sendValue)
            {
                GameAnalytics.addDesignEvent(eventId, value, fields, mergeFields);
            }
            else
            {
                GameAnalytics.addDesignEvent(eventId, fields, mergeFields);
            }

            callbackContext.success();
            return true;
        }
        else if(action.equals("addErrorEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            int severity = 0;
            String message = "";
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                severity = args.optInt("severity", 0);
                message = args.optString("message", "");
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            GameAnalytics.addErrorEvent(severity, message, fields, mergeFields);
            callbackContext.success();
            return true;
        }
        else if(action.equals("addAdEvent"))
        {
            JSONObject args = data.optJSONObject(0);
            int adAction = 0;
            int adType = 0;
            String adSdkName = "";
            String adPlacement = "";
            long duration = 0;
            boolean sendDuration = false;
            int noAdReason = 0;
            String fields = "";
            boolean mergeFields = false;

            if(args != null)
            {
                adAction = args.optInt("adAction", 0);
                adType = args.optInt("adType", 0);
                adSdkName = args.optString("adSdkName", "");
                adPlacement = args.optString("adPlacement", "");
                duration = args.optLong("duration", 0);
                sendDuration = args.has("duration");
                noAdReason = args.optInt("noAdReason", 0);
                fields = args.optString("customFields", "");
                mergeFields = args.optBoolean("mergeFields", false);
            }

            if(sendDuration)
            {
                GameAnalytics.addAdEvent(adAction, adType, adSdkName, adPlacement, duration, fields, mergeFields);
            }
            else
            {
                GameAnalytics.addAdEvent(adAction, adType, adSdkName, adPlacement, noAdReason, fields, mergeFields);
            }

            callbackContext.success();
            return true;
        }
        else if(action.equals("setEnabledInfoLog"))
        {
            boolean flag = data.optBoolean(0, false);
            GameAnalytics.setEnabledInfoLog(flag);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setEnabledVerboseLog"))
        {
            boolean flag = data.optBoolean(0, false);
            GameAnalytics.setEnabledVerboseLog(flag);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setEnabledManualSessionHandling"))
        {
            boolean flag = data.optBoolean(0, false);
            GameAnalytics.setEnabledManualSessionHandling(flag);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setEnabledEventSubmission"))
        {
            boolean flag = data.optBoolean(0, true);
            GameAnalytics.setEnabledEventSubmission(flag);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setCustomDimension01"))
        {
            String dimension = data.optString(0, "");
            GameAnalytics.setCustomDimension01(dimension);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setCustomDimension02"))
        {
            String dimension = data.optString(0, "");
            GameAnalytics.setCustomDimension02(dimension);
            callbackContext.success();
            return true;
        }
        else if(action.equals("setCustomDimension03"))
        {
            String dimension = data.optString(0, "");
            GameAnalytics.setCustomDimension03(dimension);
            callbackContext.success();
            return true;
        }
        else if (action.equals("setGlobalCustomEventFields"))
        {
            String customFields = data.optString(0, "");
            GameAnalytics.setGlobalCustomEventFields(customFields);
            callbackContext.success();
            return true;
        }
        else if(action.equals("startSession"))
        {
            GameAnalytics.startSession();
            callbackContext.success();
            return true;
        }
        else if(action.equals("endSession"))
        {
            GameAnalytics.endSession();
            callbackContext.success();
            return true;
        }
        else if(action.equals("getRemoteConfigsValueAsString"))
        {
            JSONObject args = data.optJSONObject(0);
            String key = "";
            String defaultValue = null;

            if(args != null)
            {
                key = args.optString("key", "");
                defaultValue = args.optString("defaultValue", null);
            }

            String result = null;
            if(defaultValue != null)
            {
                result = GameAnalytics.getRemoteConfigsValueAsString(key, defaultValue);
            }
            else
            {
                result = GameAnalytics.getRemoteConfigsValueAsString(key);
            }
            callbackContext.sendPluginResult(new PluginResult(Status.OK, result));
            return true;
        }
        else if(action.equals("isRemoteConfigsReady"))
        {
            boolean result = GameAnalytics.isRemoteConfigsReady();
            callbackContext.sendPluginResult(new PluginResult(Status.OK, result));
            return true;
        }
        else if(action.equals("getRemoteConfigsContentAsString"))
        {
            String result = GameAnalytics.getRemoteConfigsContentAsString();
            callbackContext.sendPluginResult(new PluginResult(Status.OK, result));
            return true;
        }
        else if(action.equals("getABTestingId"))
        {
            String result = GameAnalytics.getABTestingId();
            callbackContext.sendPluginResult(new PluginResult(Status.OK, result));
            return true;
        }
        else if(action.equals("getABTestingVariantId"))
        {
            String result = GameAnalytics.getABTestingVariantId();
            callbackContext.sendPluginResult(new PluginResult(Status.OK, result));
            return true;
        }

        return false;
    }
}

File: src/ios/GameAnalyticsCordova.m
#import "GameAnalyticsCordova.h"
#import "GameAnalytics.h"

#define VERSION @"3.2.0"

@implementation GameAnalyticsCordova

- (void)configureAvailableCustomDimensions01:(CDVInvokedUrlCommand*)command
{
    NSMutableArray* array = [command.arguments objectAtIndex:0];
    NSMutableArray *dimensions = [[NSMutableArray alloc] init];
    if(array != nil)
    {
        for (id tempObject in array)
        {
            NSString* s = [tempObject isKindOfClass:[NSString class]] ? tempObject : @"";
            [dimensions addObject:s];
        }
    }
    [GameAnalytics configureAvailableCustomDimensions01:dimensions];
}

- (void)configureAvailableCustomDimensions02:(CDVInvokedUrlCommand*)command
{
    NSMutableArray* array = [command.arguments objectAtIndex:0];
    NSMutableArray *dimensions = [[NSMutableArray alloc] init];
    if(array != nil)
    {
        for (id tempObject in array)
        {
            NSString* s = [tempObject isKindOfClass:[NSString class]] ? tempObject : @"";
            [dimensions addObject:s];
        }
    }
    [GameAnalytics configureAvailableCustomDimensions02:dimensions];
}

- (void)configureAvailableCustomDimensions03:(CDVInvokedUrlCommand*)command
{
    NSMutableArray* array = [command.arguments objectAtIndex:0];
    NSMutableArray *dimensions = [[NSMutableArray alloc] init];
    if(array != nil)
    {
        for (id tempObject in array)
        {
            NSString* s = [tempObject isKindOfClass:[NSString class]] ? tempObject : @"";
            [dimensions addObject:s];
        }
    }
    [GameAnalytics configureAvailableCustomDimensions03:dimensions];
}

- (void)configureAvailableResourceCurrencies:(CDVInvokedUrlCommand*)command
{
    NSMutableArray* array = [command.arguments objectAtIndex:0];
    NSMutableArray *currencies = [[NSMutableArray alloc] init];
    if(array != nil)
    {

        for (id tempObject in array)
        {
            NSString* s = [tempObject isKindOfClass:[NSString class]] ? tempObject : @"";
            [currencies addObject:s];
        }
    }
    [GameAnalytics configureAvailableResourceCurrencies:currencies];
}

- (void)configureAvailableResourceItemTypes:(CDVInvokedUrlCommand*)command
{
    NSMutableArray* array = [command.arguments objectAtIndex:0];
    NSMutableArray *itemTypes = [[NSMutableArray alloc] init];
    if(array != nil)
    {

        for (id tempObject in array)
        {
            NSString* s = [tempObject isKindOfClass:[NSString class]] ? tempObject : @"";
            [itemTypes addObject:s];
        }
    }
    [GameAnalytics configureAvailableResourceItemTypes:itemTypes];
}

- (void)configureBuild:(CDVInvokedUrlCommand*)command
{
    NSString* build = [command.arguments objectAtIndex:0];
    if(build == nil)
    {
        build = @"";
    }
    [GameAnalytics configureBuild:build];
}

- (void)configureAutoDetectAppVersion:(CDVInvokedUrlCommand*)command
{
    if([command.arguments objectAtIndex:0] && [[command.arguments objectAtIndex:0] isKindOfClass:[NSNumber class]] && ([[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:YES]] || [[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:NO]]))
    {
        BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
        [GameAnalytics configureAutoDetectAppVersion:flag];
    }
}

- (void)configureUserId:(CDVInvokedUrlCommand*)command
{
    NSString* uId = [command.arguments objectAtIndex:0];
    if(uId == nil)
    {
        uId = @"";
    }
    [GameAnalytics configureUserId:uId];
}

- (void)initialize:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    NSString* gameKey = nil;
    NSString* secretKey = nil;
    NSString* sdkVersion = [NSString stringWithFormat:@"cordova %@", VERSION];

    if(args != nil)
    {
        gameKey = args[@"gameKey"] && [args[@"gameKey"] isKindOfClass:[NSString class]] ? args[@"gameKey"] : @"";
        secretKey = args[@"secretKey"] && [args[@"secretKey"] isKindOfClass:[NSString class]] ? args[@"secretKey"] : @"";
        sdkVersion = args[@"sdkVersion"] && [args[@"sdkVersion"] isKindOfClass:[NSString class]] ? args[@"sdkVersion"] : [NSString stringWithFormat:@"cordova %@", VERSION];
    }

    [GameAnalytics configureSdkVersion:sdkVersion];
    [GameAnalytics initializeWithGameKey:gameKey gameSecret:secretKey];
}

- (void)addBusinessEvent:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    NSString* currency = nil;
    NSInteger amount = 0;
    NSString* itemType = nil;
    NSString* itemId = nil;
    NSString* cartType = nil;
    NSString* receipt = nil;
    BOOL autoFetchReceipt = NO;

    if(args != nil)
    {
        currency = args[@"currency"] && [args[@"currency"] isKindOfClass:[NSString class]] ? args[@"currency"] : @"";
        amount = args[@"amount"] && [args[@"amount"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"amount"] intValue] : 0;
        itemType = args[@"itemType"] && [args[@"itemType"] isKindOfClass:[NSString class]] ? args[@"itemType"] : @"";
        itemId = args[@"itemId"] && [args[@"itemId"] isKindOfClass:[NSString class]] ? args[@"itemId"] : @"";
        cartType = args[@"cartType"] && [args[@"cartType"] isKindOfClass:[NSString class]] ? args[@"cartType"] : @"";
        receipt = args[@"receipt"] && [args[@"cartType"] isKindOfClass:[NSString class]] ? args[@"cartType"] : @"";
        autoFetchReceipt = args[@"autoFetchReceipt"] && [args[@"autoFetchReceipt"] isKindOfClass:[NSNumber class]] && [args[@"autoFetchReceipt"] isEqual:[NSNumber numberWithBool:YES]] ? [[args valueForKey:@"autoFetchReceipt"] boolValue] : NO;
    }

    if(autoFetchReceipt)
    {
        [GameAnalytics addBusinessEventWithCurrency:currency amount:amount itemType:itemType itemId:itemId cartType:cartType autoFetchReceipt:autoFetchReceipt];
    }
    else
    {
        [GameAnalytics addBusinessEventWithCurrency:currency amount:amount itemType:itemType itemId:itemId cartType:cartType receipt:receipt];
    }
}

- (void)addResourceEvent:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    int flowType = 0;
    NSString* currency = nil;
    double amount = 0;
    NSString* itemType = nil;
    NSString* itemId = nil;

    if(args != nil)
    {
        flowType = args[@"flowType"] ? [[args valueForKey:@"flowType"] intValue] : 0;
        currency = args[@"currency"];
        amount = args[@"amount"] && [args[@"amount"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"amount"] doubleValue] : 0;
        itemType = args[@"itemType"];
        itemId = args[@"itemId"];
    }

    [GameAnalytics addResourceEventWithFlowType:(GAResourceFlowType)flowType currency:currency amount:@(amount) itemType:itemType itemId:itemId];
}

- (void)addProgressionEvent:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    int progressionStatus = 0;
    NSString* progression01 = nil;
    NSString* progression02 = nil;
    NSString* progression03 = nil;
    int score = 0;
    BOOL sendScore = NO;

    if(args != nil)
    {
        progressionStatus = args[@"progressionStatus"] && [args[@"progressionStatus"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"progressionStatus"] intValue] : 0;
        progression01 = args[@"progression01"] && [args[@"progression01"] isKindOfClass:[NSString class]] ? args[@"progression01"] : @"";
        progression02 = args[@"progression02"] && [args[@"progression02"] isKindOfClass:[NSString class]] ? args[@"progression02"] : @"";
        progression03 = args[@"progression03"] && [args[@"progression03"] isKindOfClass:[NSString class]] ? args[@"progression03"] : @"";
        score = args[@"score"] && [args[@"score"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"score"] intValue] : 0;
        sendScore = args[@"score"] && [args[@"score"] isKindOfClass:[NSNumber class]] ? YES : NO;
    }

    if(sendScore)
    {
        [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus progression01:progression01 progression02:progression02 progression03:progression03 score:score];
    }
    else
    {
        [GameAnalytics addProgressionEventWithProgressionStatus:(GAProgressionStatus)progressionStatus progression01:progression01 progression02:progression02 progression03:progression03];
    }
}

- (void)addDesignEvent:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    NSString* eventId = nil;
    double value = 0;
    BOOL sendValue = NO;

    if(args != nil)
    {
        eventId = args[@"eventId"] && [args[@"eventId"] isKindOfClass:[NSString class]] ? args[@"eventId"] : @"";
        value = args[@"value"] && [args[@"value"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"value"] doubleValue] : 0;
        sendValue = args[@"value"] && [args[@"value"] isKindOfClass:[NSNumber class]] ? YES : NO;
    }

    if(sendValue)
    {
        [GameAnalytics addDesignEventWithEventId:eventId value:@(value)];
    }
    else
    {
        [GameAnalytics addDesignEventWithEventId:eventId];
    }
}

- (void)addErrorEvent:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];

    int severity = 0;
    NSString* message = nil;

    if(args != nil)
    {
        severity = args[@"severity"] && [args[@"severity"] isKindOfClass:[NSNumber class]] ? [[args valueForKey:@"severity"] intValue] : 0;
        message = args[@"message"] && [args[@"message"] isKindOfClass:[NSString class]] ? args[@"message"] : @"";
    }

    [GameAnalytics addErrorEventWithSeverity:(GAErrorSeverity)severity message:message];
}

- (void)setEnabledInfoLog:(CDVInvokedUrlCommand*)command
{
    if([command.arguments objectAtIndex:0] && [[command.arguments objectAtIndex:0] isKindOfClass:[NSNumber class]] && ([[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:YES]] || [[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:NO]]))
    {
        BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
        [GameAnalytics setEnabledInfoLog:flag];
    }
}

- (void)setEnabledVerboseLog:(CDVInvokedUrlCommand*)command
{
    if([command.arguments objectAtIndex:0] && [[command.arguments objectAtIndex:0] isKindOfClass:[NSNumber class]] && ([[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:YES]] || [[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:NO]]))
    {
        BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
        [GameAnalytics setEnabledVerboseLog:flag];
    }
}

- (void)setEnabledManualSessionHandling:(CDVInvokedUrlCommand*)command
{
    if([command.arguments objectAtIndex:0] && [[command.arguments objectAtIndex:0] isKindOfClass:[NSNumber class]] && ([[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:YES]] || [[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:NO]]))
    {
        BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
        [GameAnalytics setEnabledManualSessionHandling:flag];
    }
}

- (void)setEnabledEventSubmission:(CDVInvokedUrlCommand*)command
{
    if([command.arguments objectAtIndex:0] && [[command.arguments objectAtIndex:0] isKindOfClass:[NSNumber class]] && ([[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:YES]] || [[command.arguments objectAtIndex:0] isEqual:[NSNumber numberWithBool:NO]]))
    {
        BOOL flag = [[command.arguments objectAtIndex:0] boolValue];
        [GameAnalytics setEnabledEventSubmission:flag];
    }
}

- (void)setCustomDimension01:(CDVInvokedUrlCommand*)command
{
    NSString* dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension01:dimension];
}

- (void)setCustomDimension02:(CDVInvokedUrlCommand*)command
{
    NSString* dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension02:dimension];
}

- (void)setCustomDimension03:(CDVInvokedUrlCommand*)command
{
    NSString* dimension = [command.arguments objectAtIndex:0];
    [GameAnalytics setCustomDimension03:dimension];
}

- (void)startSession:(CDVInvokedUrlCommand*)command
{
    [GameAnalytics startSession];
}

- (void)endSession:(CDVInvokedUrlCommand*)command
{
    [GameAnalytics endSession];
}

- (void)getRemoteConfigsValueAsString:(CDVInvokedUrlCommand*)command
{
    NSMutableDictionary* args = [command.arguments objectAtIndex:0];
    NSString* key = nil;
    NSString* defaultValue = nil;

    if(args != nil)
    {
        key = args[@"key"] && [args[@"key"] isKindOfClass:[NSString class]] ? args[@"key"] : @"";
        defaultValue = args[@"defaultValue"] && [args[@"defaultValue"] isKindOfClass:[NSString class]] ? args[@"defaultValue"] : @"";
    }

    NSString* result = nil;
    if(defaultValue)
    {
        result = [GameAnalytics getRemoteConfigsValueAsString:key defaultValue:defaultValue];
    }
    else
    {
        result = [GameAnalytics getRemoteConfigsValueAsString:key];
    }
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:result];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)isRemoteConfigsReady:(CDVInvokedUrlCommand*)command
{
    BOOL result = [GameAnalytics isRemoteConfigsReady];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:result];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getRemoteConfigsContentAsString:(CDVInvokedUrlCommand*)command
{
    NSString* result = [GameAnalytics getRemoteConfigsContentAsString];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:result];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getABTestingId:(CDVInvokedUrlCommand*)command
{
    NSString* result = [GameAnalytics getABTestingId];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:result];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getABTestingVariantId:(CDVInvokedUrlCommand*)command
{
    NSString* result = [GameAnalytics getABTestingVariantId];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:result];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end
